Домашнее задание №2

IlyatTsk opened the pull request:
1. Создан класс EndpointConfig
2. В пакете clients созданы необходимые классы с описанными в них методами Actions и Crud
3. Был создан вспомогательный класс Duck, содержащий необходимые поля для использования в методе create

IlyatTsk committed:
1.	Создан класс EndpointConfig
2.	В пакете clients созданы необходимые классы с описанными в них методами Actions и Crud
3. Содзан класс Duck для создания уточки с необходимыми полями для метода create

shalyaa commented on src/test/java/clients/DuckCRUDClient.java:
```diff
@@ -0,0 +1,60 @@
+    public void duckCreate(TestCaseRunner runner,Duck duck) throws JsonProcessingException {
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        String jsonDuck = objectMapper.writeValueAsString(duck);
```
хорошее решение использовать дополнительный класс для описания тела запроса и преобраовывать все в json, но на 22 занятии расскажу, как сделать все куда проще с помощью дополнительной библиотеки =)

shalyaa commented on src/test/java/models/Duck.java:
```diff
@@ -0,0 +1,58 @@
+        return height;
+    }
+
+    public void setHeight(int height) {
```
Если посмотреть внимательнее на сваггер, то поле height имеет тип double

shalyaa replied:
поменяй тип переменной на double и можно сливать в мастер =)

shalyaa commented on src/test/java/tests/EndpointConfig.java:
```diff
@@ -0,0 +1,25 @@
+
+public class EndpointConfig {
+
+    public static final String CREATE = "/api/duck/create";
```
вынести эндпоинты в отдельные переменные решение интересное, и для этого проекта подойдет, но имей ввиду, что если АПИ-сервис будет большой, к примеру в нем будет 30+ эндпоинов или проект будет использоваться не для одного АПИ-сервиса, а для нескольких, то такое решение не подойдет и эндпоинты придется передавать все таки в классах папки client

IlyatTsk committed:
1.	Исправил тип переменной height с int на double
2.	Создал для класса DuckActionsClient и DuckCRUDClient по 3 метода валидации:
a.	С передачей ответа String’ой
b.	С передачей ответа из папки Resources
c.	С передачей ответа из папки Payload
3.	Создал папку payloads, в ней описал классы модели для запроса POST /api/duck/create и всех эндпоинтов с учетом структуры ответа

shalyaa commented on src/test/java/clients/DuckActionsClient.java:
```diff
@@ -20,6 +26,35 @@ public void duckFly(TestCaseRunner runner, String id) {
     }
 
+    @Description("Валидация полученного ответа String'ой метода Fly /api/duck/action/fly")
+    public void validateResponseWithString(TestCaseRunner runner, String expectedFile) {
```
Вынести ожидаемое тело ответа в переменную правильное решение, но не вижу, чтобы ты использовать эту переменную в body
так же переменную лучше назвать понятнее, например expectedBody или expectedString, так как никакого файла ты в этом методе не передаешь

shalyaa commented on src/test/java/clients/DuckCRUDClient.java:
```diff
@@ -32,6 +36,39 @@ public void duckCreate(TestCaseRunner runner,Duck duck) throws JsonProcessingExc
     }
 
+    @Description("Валидация полученного ответа String'ой метода Create /api/duck/create")
+    public void validateResponseWithString(TestCaseRunner runner, String expectedFile) {
```
такая же ошибка, как и в другом классе, переменную сделал, но не используешь ее

shalyaa commented on src/test/java/models/Duck.java:
```diff
@@ -3,7 +3,7 @@
 public class Duck {
```
класс Duck нам не понадобится в тестах - можно удлаить

shalyaa commented on src/test/java/payloads/EmptyResponse.java:
```diff
@@ -0,0 +1,11 @@
+package payloads;
```
Пустой ответ описывать не нужно, в тесте достаточно будет передать стрингой ответ, так что можно удалить класс этот

shalyaa replied:
но классно, что подумал об описании и пустого ответа))

shalyaa commented on src/test/java/payloads/IntArrayResponse.java:
```diff
@@ -0,0 +1,15 @@
+@Setter
+@Accessors(fluent = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class IntArrayResponse {
```
у нас разве есть в ответе или запросе утки json с передачей id?

shalyaa commented on src/test/java/payloads/PostRequestCreate.java:
```diff
@@ -0,0 +1,27 @@
+@Setter
+@Accessors(fluent = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class PostRequestCreate {
```
Посмотри внимательно на запрос /api/duck/create, там есть еще одно поле в ответе, тогда ты сможешь использовать этот класс для создания тела и запроса и ответа создания утки

shalyaa commented on src/test/java/payloads/StringMessageResponse.java:
```diff
@@ -0,0 +1,15 @@
+
+@Setter
+@Accessors(fluent = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
нам точно нужна эта аннотация. с учетом, что тут только 1 поле?

shalyaa commented on src/test/java/clients/DuckCRUDClient.java:
```diff
@@ -19,7 +23,7 @@ public class DuckCRUDClient {
-    public void duckCreate(TestCaseRunner runner,Duck duck) throws JsonProcessingException {
+    public void duckCreate(TestCaseRunner runner, Duck duck) throws JsonProcessingException {
 
         ObjectMapper objectMapper = new ObjectMapper();
```
ObjectMapper нам не понадобится, так что можно удалить всю эту конструкцию, вообще на будущее - все лишнее смело удаляем из проекта, при желании всегда можно по коммитам найти удаленное и вернуть)
